tracer/src/Datadog.Trace/ClrProfiler/AutomaticTracer.cs

```diff
@@ -40,6 +52,15 @@ SpanContext IDistributedTracer.GetSpanContext()
             return SpanContextPropagator.Instance.Extract(value);
         }
 
+        void IDistributedTracer.SetSpanContextRaw(IReadOnlyDictionary<string, string> value)
+        {
+            // This is a performance optimization. See comment in GetDistributedTrace() about potential race condition
+            if (_child != null)
+            {
+                DistributedTrace.Value = value;
+            }
+        }
+
         void IDistributedTracer.SetSpanContext(SpanContext value)
```

Comment:
Since we haven't shipped this yet, I was thinking we could refactor this method to take the `IReadonlyDictionary<string, string>` instead of having a new `SetSpanContextRaw` API. What do you think?

Reply:
Makes sense

Reply:
I still haven't seen all the code involved in the version mismatch fix, but for future-proofing purposes I think we should use `IReadOnlyDictionary<string, string>` (or something like it) everywhere for propagation rather than `SpanContext`. The latter is a public type that can be hard to modify (especially `ISpanContext`), and it doesn't (and _shouldn't_) contain all the values we propagate.

Propagated values include (grouped into logical "levels"):

- span level
  - span id 
- local trace level
  - sampling priority (which today is awkwardly spread between `SpanContext` and `TraceContext`)
- distributed trace level
  - trace id
  - origin (lives in `SpanContext` today)
  - internal Datadog tags (I am adding to `TraceContext` now)
    - list of upstream services (service name, sampling priority, sampling mechanism, sampling rate)
    - (future tags)
  - baggage (future feature)

This also relates to my comments earlier today about refactoring the object model to make this more... sane.

Reply:
@lucaspimentel this is sort of a crystal ball question, but do you anticipate any issues with the plan to only store strings as values?

Reply:
Updated in c8ac3e66e537a5acad59038f9c19fdd3fd206f40

Reply:
> do you anticipate any issues with the plan to only store strings as values?

If you mean whether we may have data that's not easily converted to/from string, then not really. Anything that's propagated is already required to be converted to/from ASCII strings for http headers. Any new tags will be designed with that in mind.

I have some (small-ish) concerns about the overhead of converting ints and floats back and forth, but we may be able to cache some of those.

Reply:
> If you mean whether we may have data that's not easily converted to/from string, then not really.

That was my main concern, thanks for confirming